### PR TITLE
Build for Scala 3 (final)

### LINKED ISSUE #257: Scala 3 support
Hi! I would be interested in bringing Odin to Scala 3.0.0-RC1, do you see any potential blockers, missing dependencies, macros or anything else that could be an issue?

### PR DESCRIPTION
**Pull request by ybasket**
Scala 3.0.0 cross build for all Odin modules. Some smaller problems still to be tackled or documented:
- [x]  Document limitations of magnolia on Scala 3 (value classes, maybe more)
- [x]  Benchmarks don't compile due to some weird error with a type from scribe
- [x]  ~Clarify `FunctionK` usage on Scala 3 (cats seems to have nothing yet to provide source compatibility for `FunctionK.lift` and friends)~ ~No blocker, only a small beautification in a single place (zio module). Will check on typelevel discord and open a follow-up PR if necessary.~ Update: There some PRs for cats, but there are some compatibility problems, so it'll probably take a while. Odin's use of it is very little and the workaround just two extra lines, so I suggest to ignore for now and optimise when cats is ready.

Includes update for all library dependencies to their latest versions.

Closes #257

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -136,7 +147,16 @@ lazy val `odin-slf4j` = (project in file("slf4j"))
 lazy val `odin-extras` = (project in file("extras"))
   .settings(sharedSettings)
   .settings(
-    libraryDependencies += magnolia
+    libraryDependencies ++= (CrossVersion.partialVersion(scalaVersion.value) match {
+      case Some((3, _)) => List(magnoliaScala3)
+      case _ =>
+        List(
+          magnoliaScala2,
+          // only in provided scope so that users of extras not relying on magnolia don't get it on their classpaths
```
Comment (ybasket) [inline build.sbt:155]: I guess this was implicitly already the case before? Thoughts?

File: build.sbt [hunk 2]
```diff
@@ -174,42 +188,44 @@ lazy val examples = (project in file("examples"))
...
+  .dependsOn(`odin-core`, `odin-json`, `odin-zio`, /* `odin-monix`,*/ `odin-slf4j`, `odin-extras`)
+  .aggregate(`odin-core`, `odin-json`, `odin-zio`, /* `odin-monix`,*/ `odin-slf4j`, `odin-extras`, benchmarks, examples)
+
+def scalacOptionsVersion(scalaVersion: String) = CrossVersion.partialVersion(scalaVersion) match {
+  case Some((2, scalaMajor)) if scalaMajor == 12 => scalac2Options ++ scalac212Options
+  case Some((2, scalaMajor)) if scalaMajor == 13 => scalac2Options ++ scalac213Options
+  case Some((3, _))                              => scalac3Options
+}
+
+lazy val scalac2Options = Seq(
```
[the remark this answers is not in the document]
Comment (sergeykolbasov) [reply build.sbt:206]: I prefer to have them explicit, as I sometimes shut down/adjust them if needed

File: core/src/main/scala/io/odin/formatter/Formatter.scala [hunk 3]
```diff
@@ -70,49 +73,45 @@ object Formatter {
...
-  /**
-    * Default Throwable printer is twice as slow. This method was borrowed from scribe library.
+  /** Default Throwable printer is twice as slow. This method was borrowed from scribe library.
     *
-    * The result differs depending on the format:
-    * `ThrowableFormat.Depth.Full` - prints all elements of a stack trace
-    * `ThrowableFormat.Depth.Fixed` - prints N elements of a stack trace
-    * `ThrowableFormat.Indent.NoIndent` - prints a stack trace without indentation
-    * `ThrowableFormat.Indent.Fixed` - prints a stack trace prepending every line with N spaces
+    * The result differs depending on the format: `ThrowableFormat.Depth.Full` - prints all elements of a stack trace
```
Comment (sergeykolbasov) [inline core/src/main/scala/io/odin/formatter/Formatter.scala:111]: something is awfully off with formatting of all the docs

Comment (ybasket) [reply core/src/main/scala/io/odin/formatter/Formatter.scala]: @sergeykolbasov I fixed the formatting of the docs, but by updating `scalafmt` (skipping many versions to the most recent 3.0.0-RC4 as it has best Scala 3 support), there are some changes that seem unavoidable – they happen on `master` as well. If that's an issue for getting this PR through, we can  update `scalafmt` in a separate PR first (defining the style there) and then the diff of this PR should be more focused on the actual code changes for Scala 3. What do you prefer?

Comment (sergeykolbasov) [reply core/src/main/scala/io/odin/formatter/Formatter.scala]: Hmm, I don't mind style-driven commits as far as it's tooling-based, but I think `scalafmt` did something wrong in general. I'll check their latest release to see if there is a bug or configuration missing, but breaking docs newlines is definitely something that the formatting tools shouldn't do.

Comment (sergeykolbasov) [reply core/src/main/scala/io/odin/formatter/Formatter.scala]: ah, you already did

nice

File: README.md [hunk 4]
```diff
@@ -683,7 +683,12 @@ service.findAndVerify("bad-user").attempt.unsafeRunSync()
 
 ### Extras. Derivation
 
-`io.odin.extras.derivation.render` provides a Magnolia-based derivation of the `Render` type class.
```
Comment (sergeykolbasov) [inline README.md:686]: Those changes will be lost on release. Could you change `docs/README.md` instead?

**Comment by sergeykolbasov** [comment]
@ybasket Thanks for the awesome work! It looks good to me. Could you address minor documentation comment and fix the conflicting files? Then we're good to merge

**Comment by ybasket** [comment]
> @ybasket Thanks for the awesome work! It looks good to me. Could you address minor documentation comment and fix the conflicting files? Then we're good to merge

@sergeykolbasov You're welcome. Thanks for developing and maintaining Odin in the first place!

Fixed the README issue and resolved conflicts, so if CI agrees, we should be good now.

**Comment by sergeykolbasov** [comment]
Hi @ybasket 

There are issues with scoverage & Scala 3. I'm looking into it right now. I'll merge your branch as is and then apply my fix on top

**Pull request merged**